### PR TITLE
Adding support for BNO055

### PR DESCRIPTION
**Pull request by Ellerbach**
Adding support for BNO055
* Full implementation of sensors setup
* Full implementation of sensor reading
* Full implementation of units selection
* Full implementation of calibration with example
* Full implementation of power modes
* Full implementation of sensor selection
* Full implementation of axis remap for accelerator
* Documentation

### CODE REVIEW

File: src/devices/Bno055/Bno055Sensor.cs [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+{
+    public class Bno055Sensor : IDisposable
+    {
+        /// <summary>
+        /// The default I2C Address
+        /// </summary>
+        public const byte DefaultI2cAddress = 0x28;
+
+        /// <summary>
+        /// If setup this is a valid I2C Address as well
```
Comment (krwq) [inline src/devices/Bno055/Bno055Sensor.cs]: comment might need some re-wording :smile:

File: src/devices/Bno055/Bno055Sensor.cs [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+            axisSettings[2].Axis = (AxisMap)((retMap >> 4) & 0x03);
+            axisSettings[0].Sign = (AxisSign)(retSign & 0x01);
+            axisSettings[1].Sign = (AxisSign)((retSign >> 1) & 0x01);
+            axisSettings[2].Sign = (AxisSign)((retSign >> 2) & 0x01);
+            SetConfigMode(false);
+            return axisSettings;
+        }
+
+        /// <summary>
+        /// Get the orientation (Eular Angles) X = Heading, Y = Roll, Z = Pitch
```
Comment (krwq) [inline src/devices/Bno055/Bno055Sensor.cs]: Euler

File: src/devices/Bno055/Bno055Sensor.cs [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+        /// </summary>
+        public Vector3 Orientation
+        {
+            get
+            {
+                var retVect = GetVectorData(Registers.EULER_H_LSB);
+                // If unit is MeterG, then divide by 900, otherwise divide by 16
+                if ((_units & Units.EulerAnglesRadians) == Units.EulerAnglesRadians)
```
Comment (krwq) [inline src/devices/Bno055/Bno055Sensor.cs]: I think you can use HasFlag here

File: src/devices/Bno055/Bno055Sensor.cs [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+                    return retVect / 900;
+                else
+                    return retVect / 16;
+            }
+        }
+
+        /// <summary>
+        /// Get the accelerometer
+        /// </summary>
+        public Vector3 Accelerometer
```
Comment (krwq) [inline src/devices/Bno055/Bno055Sensor.cs]: What's the difference between this acceleration and the LinearAcceleration?

File: src/devices/Bno055/Bno055Sensor.cs [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+                if ((_units & Units.AccelerationMeterG) == Units.AccelerationMeterG)
+                    return retVect;
+                else
+                    return retVect / 100;
+            }
+        }
+        /// <summary>
+        /// Get the gravity
+        /// </summary>
+        public Vector3 Gravity
```
Comment (krwq) [inline src/devices/Bno055/Bno055Sensor.cs]: Another acceleration? Now I'm confused...

File: src/devices/Bno055/Models/AxisRemap.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    /// The dot is the one on the chip
+    ///                    | Z axis
+    ///                    |
+    ///                    |   / X axis
+    ///                ____|__/____
+    ///   Y axis     / *   | /    /|
+    ///   _________ /______|/    //
+    ///            /___________ //
+    ///            |____________|/
+    /// </summary>
```
Comment (krwq) [inline src/devices/Bno055/Models/AxisRemap.cs]: nice ascii art

Comment (ZhangGaoxing) [reply src/devices/Bno055/Models/AxisRemap.cs]: unbelievable 👍👍👍

File: src/devices/Bno055/Bno055Sensor.cs (hunk 4 above)
Comment (Ellerbach) [reply src/devices/Bno055/Bno055Sensor.cs]: I explained all in the documentation. I'll copy/paste in the code as well

File: src/devices/Bno055/Bno055Sensor.cs (hunk 5 above)
Comment (Ellerbach) [reply src/devices/Bno055/Bno055Sensor.cs]: I explained in the documentation. In short: Accelerometer = Gravity + Linear Acceleration.
I've Added more details in the comments.

File: src/devices/Bno055/samples/README.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+# Bno055 sensor example
+
+Please refer to [main documentation](../README.md) for more details.
+
+When using this sensor with an I2C board, just make sure you connect your I2C pins correctly as well as the alimentation adn ground.
```
Comment (krwq) [inline src/devices/Bno055/samples/README.md]: what do you mean by `alimentation`? Also type in `and` (second word from the last)

File: src/devices/Bno055/Bno055Sensor.cs [hunk 8]
```diff
@@ -0,0 +1,495 @@
...
+
+using System;
+using System.Buffers.Binary;
+using System.Device.I2c;
+using System.Numerics;
+using System.Threading;
+
+namespace Iot.Device.Bno055
+{
+    public class Bno055Sensor : IDisposable
```
Comment (joshfree) [inline src/devices/Bno055/Bno055Sensor.cs]: could you add a link to the doc/pdf where the addresses came from?

File: src/devices/Bno055/Bno055Sensor.cs [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+        /// </summary>
+        public Info Info { get; internal set; }
+
+        /// <summary>
+        /// Create an BNO055 sensor
+        /// </summary>
+        /// <param name="i2cDevice">The I2C Device</param>
+        /// <param name="operationMode">The operation mode to setup</param>
+        /// <param name="autoDispoable">true to dispose the I2C device at dispose</param>
+        public Bno055Sensor(I2cDevice i2cDevice, OperationMode operationMode = OperationMode.AccelerometerMagnetometerGyroscopeRelativeOrientation, bool autoDispoable = true)
```
Comment (joshfree) [inline src/devices/Bno055/Bno055Sensor.cs]: > AccelerometerMagnetometerGyroscopeRelativeOrientation

What a mouthful 😄

Comment (Ellerbach) [reply src/devices/Bno055/Bno055Sensor.cs]: You know, sensors where you can activate and deactivate some of the elements -:) And I know you guys prefer when it's explicit :-) Can't do longer :-)

**Comment by krwq** [comment]
Thanks @Ellerbach!

**Pull request merged**